### DISCUSSION TITLE
How can I avoid duplication of code?

### DISCUSSION THREAD

**Discussion by ughstudios** [aio-libs/aiohttp#6793] [Q&A]
```
async def get_active_player_stats() -> Optional[ActivePlayer]:
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get('https://127.0.0.1:2999/liveclientdata/activeplayer') as response:
                active_player_json = await response.json()
                return ActivePlayer.parse_obj(active_player_json)
    except ValidationError as exc:
        logger.debug(exc)
    return None


async def get_game_events() -> Optional[Events]:
    login_token = get_login_token()
    if not login_token:
        return None

    app_port = get_app_port()
    if not app_port:
        return None

    async with aiohttp.ClientSession() as session:
        async with session.get('https://127.0.0.1:2999/liveclientdata/eventdata') as response:
            json = await response.json()
            return Events.parse_obj(json)
```

`async with aiohttp.ClientSession() as session:` is duplicated every time I want to make an http request with aiohttp

So I have the following code, and I am going to be making multiple get requests using this aiohttp. How can I avoid having to copy and paste this multiple times throughout my code? I cannot cache it with a memoize decorator. I cannot cache it because it's `async`, so how do I cache a `ClientSession`? I would like to have one `ClientSession` object that is used everywhere throughout my code.

**Comment by Dreamsorcerer** [comment]
I don't know what your core application/framework looks like, but if you were building an aiohttp (server) application, you could just put it in a cleanup context like:

```
async def client_session(app):
    async with ClientSession() as session:
        app["client"] = session
        yield
app.cleanup_ctx.append(client_session)
```

I'm sure you could do something similar to ensure it gets cleaned up when your application shuts down.

Also, you don't _need_ to use `async with`, it just ensures auto-closing, much like file objects and such. You can instead skip the context manager and ensure you explicitly run `await session.close()` when the session is finished with. You can see in the code, that is literally all the context manager does: https://github.com/aio-libs/aiohttp/blob/master/aiohttp/client.py#L1070-L1079
